File: Python/ByRound/1387/1387_A_Graph.py
# Contest 1387, Problem A: Graph
# URL: https://codeforces.com/contest/1387/problem/A

from sys import stdin
input=stdin.readline
n,m=map(int,input().split())
g=[[] for _ in range(n)]
for _ in range(m):
    u,v,w=map(int,input().split())
    u-=1
    v-=1
    g[u].append((w*2,v))
    g[v].append((w*2,u))
vis=[0]*n
val=[0]*n
k=[0]*n
b=[0]*n
for i in range(n):
    if vis[i]:
        continue
    q=[i]
    vis[i]=1
    x=0
    nx=0
    k[i]=1
    t=0
    while t<len(q):
        u=q[t]
        t+=1
        for w,v in g[u]:
            if not vis[v]:
                vis[v]=1
                k[v]=-k[u]
                b[v]=w-b[u]
                q.append(v)
            elif nx:
                if k[u]*x+b[u]+k[v]*x+b[v]!=w:
                    print("NO")
                    exit()
            else:
                if k[u]==-k[v]:
                    if b[u]+b[v]==w:
                        continue
                    else:
                        print("NO")
                        exit()
                nx=1
                x=(w-b[u]-b[v])//(k[u]+k[v])
    if not nx:
        ps=[-k[i]*b[i] for i in q]
        ps.sort()
        x=ps[len(ps)//2]
    for i in q:
        val[i]=k[i]*x+b[i]
print("YES")
print(*[i/2 for i in val])

File: CPP/ByRound/1387/1387_A_Graph.cpp
/*
 * Problem URL : https://codeforces.com/contest/1387/problem/A
 * Submit Date : 2025-09-02
 */

#include <bits/stdc++.h>
using namespace std;
#define pii pair<int, int>
#define fi first
#define se second
#define rep(i, j, k) for (int i = (j); i <= (k); i++)
#define per(i, j, k) for (int i = (j); i >= (k); i--)
const int N = 114514;
vector<pii> e[N];
int n, m, a[N], k[N], g[N], v[N], s[N], as[N], fl, po, x, y, z, kx;
void dfs(int x) {
  v[x] = 1;
  s[++po] = x;
  int y;
  for (pii w : e[x])
    if (v[y = w.fi]) {
      if (a[x] + a[y] == 0 && k[x] + k[y] != w.se)
        puts("NO"), exit(0);
      if (a[x] == a[y]) {
        int nk = (w.se - k[x] - k[y]) / (a[x] + a[y]);
        if (fl && nk != kx)
          puts("NO"), exit(0);
        else
          fl = 1, kx = nk;
      }
    } else
      a[y] = -a[x], k[y] = w.se - k[x], dfs(y);
}
int main() {
  scanf("%d%d", &n, &m);
  rep(i, 1, m) scanf("%d%d%d", &x, &y, &z), z *= 2, e[x].push_back({y, z}),
      e[y].push_back({x, z});
  rep(i, 1, n) if (!v[i]) {
    a[i] = 1;
    kx = k[i] = fl = po = 0;
    dfs(i);
    if (!kx) {
      rep(i, 1, po) g[i] = -k[s[i]] / a[s[i]];
      sort(g + 1, g + po + 1);
      kx = g[(po + 1) / 2];
    }
    rep(i, 1, po) as[s[i]] = kx * a[s[i]] + k[s[i]];
  }
  puts("YES");
  rep(i, 1, n) printf("%.2lf ", 0.5 * as[i]);
}
